src/affine3d.rs

```diff
@@ -591,59 +489,56 @@ impl core::ops::Mul for Affine3D {
-            let out_x_row = _mm_add_ps(_mm_add_ps(v_x, v_z), _mm_add_ps(v_y, v_w));
+            let v_x = _mm_mul_ps(rhs_x_col.splat_x(), lhs_x_col);
+            let v_y = _mm_mul_ps(rhs_x_col.splat_y(), lhs_y_col);
+            let v_z = _mm_mul_ps(rhs_x_col.splat_z(), lhs_z_col);
+            let out_x_col = _mm_add_ps(_mm_add_ps(v_x, v_z), v_y);
+
+            let v_x = _mm_mul_ps(rhs_y_col.splat_x(), lhs_x_col);
+            let v_y = _mm_mul_ps(rhs_y_col.splat_y(), lhs_y_col);
+            let v_z = _mm_mul_ps(rhs_y_col.splat_z(), lhs_z_col);
+            let out_y_col = _mm_add_ps(_mm_add_ps(v_x, v_z), v_y);
 
-            let v_x = _mm_mul_ps(lhs_y_row.splat_x(), rhs_x_row);
-            let v_y = _mm_mul_ps(lhs_y_row.splat_y(), rhs_y_row);
-            let v_z = _mm_mul_ps(lhs_y_row.splat_z(), rhs_z_row);
-            let v_w = _mm_mul_ps(lhs_y_row.splat_w(), rhs_w_row);
-            let out_y_row = _mm_add_ps(_mm_add_ps(v_x, v_z), _mm_add_ps(v_y, v_w));
+            let v_x = _mm_mul_ps(rhs_z_col.splat_x(), lhs_x_col);
+            let v_y = _mm_mul_ps(rhs_z_col.splat_y(), lhs_y_col);
+            let v_z = _mm_mul_ps(rhs_z_col.splat_z(), lhs_z_col);
+            let out_z_col = _mm_add_ps(_mm_add_ps(v_x, v_z), v_y);
 
-            let v_x = _mm_mul_ps(lhs_z_row.splat_x(), rhs_x_row);
-            let v_y = _mm_mul_ps(lhs_z_row.splat_y(), rhs_y_row);
-            let v_z = _mm_mul_ps(lhs_z_row.splat_z(), rhs_z_row);
-            let v_w = _mm_mul_ps(lhs_z_row.splat_w(), rhs_w_row);
-            let out_z_row = _mm_add_ps(_mm_add_ps(v_x, v_z), _mm_add_ps(v_y, v_w));
+            let v_x = _mm_mul_ps(rhs_w_col.splat_x(), lhs_x_col);
+            let v_y = _mm_mul_ps(rhs_w_col.splat_y(), lhs_y_col);
+            let v_z = _mm_mul_ps(rhs_w_col.splat_z(), lhs_z_col);
+            let out_w_col = _mm_add_ps(_mm_add_ps(v_x, v_z), _mm_add_ps(v_y, lhs_w_col));
 
             Self {
-                x_row: out_x_row.into(),
-                y_row: out_y_row.into(),
-                z_row: out_z_row.into(),
+                x_col: out_x_col.into(),
+                y_col: out_y_col.into(),
+                z_col: out_z_col.into(),
+                w_col: out_w_col.into(),
             }
         }
     }
 
-    #[cfg(target_arch = "spirv")]
-    #[inline(always)]
-    fn mul(self, rhs: Affine3D) -> Self::Output {
-        Affine3D::from_mat4(Mat4::from(self) * Mat4::from(rhs))
-    }
-
-    #[cfg(not(any(
-        target_arch = "spirv",
-        all(target_feature = "sse2", not(feature = "scalar-math"))
-    )))]
+    #[cfg(not(all(target_feature = "sse2", not(feature = "scalar-math"))))]
     #[inline(always)]
     fn mul(self, rhs: Affine3D) -> Self::Output {
         Self {
-            x_row: rhs.mul_row_vec4_from_left(self.x_row),
-            y_row: rhs.mul_row_vec4_from_left(self.y_row),
-            z_row: rhs.mul_row_vec4_from_left(self.z_row),
+            x_col: self.transform_vector3(rhs.x_col.into()).into(),
```

Comment:
These are going from `Vec3A` to `Vec3` then effectively splatted x, y, and z. Could probably be made faster if you added `transform_vector3a` which is the actual implementation and keeps everything in SIMD, avoids storing the SIMD columns into 3xf32. Same goes for the call to `transform_point3`.

You might be able to get parity with `Affine3D * Affine3D` doing that.

Reply:
This is the scalar path, so sticking to Vec3A is actually slower (I tested it).

Reply:
Ah my bad, didn't notice it was the scalar path.